memory_reference/src/algo.hpp

```diff
@@ -0,0 +1,118 @@
+#include <pybind11/numpy.h>
+#include <pybind11/stl.h>
+#include "tags.hpp"
+#include "mem_ref_detail.hpp"
+#include "cupy_ref.hpp"
+#include "cupy_caster.hpp"
+#include "cupy_allocate.hpp"
+
+template<typename TDevice>
+class Algo {
+    public:
+        void whoami();
+
+        Algo();
+        void initialize_array(int size);
+        Mem_Ref<TDevice> get_input_memory();
+        void compute(Mem_Ref<TDevice> input, Mem_Ref<TDevice> output);
+        Mem_Ref<TDevice> get_ouput_memory();
+};
+
+template<>
+class Algo<CPU> {
+public:
+    Mem_Ref<CPU> input;
+    Mem_Ref<CPU> output;
+
+    int size;
+
+    void whoami()
+    {
+        std::cout << "I'm the CPU version\n";
+    }
+
+    void initialize_array(int size)
+    {
+        this->size = size;
+
+        input = Mem_Ref<CPU>(size);
+        pybind11::buffer_info bufInput = input.request();
+        double *ptrInput = static_cast<double*>(bufInput.ptr);
+        
+        output = Mem_Ref<CPU>(size);
+        pybind11::buffer_info bufOutput = output.request();
+        double *ptrOutput = static_cast<double*>(bufOutput.ptr);
+
+        for(int i = 0; i < size; i++)
+        {
+            ptrOutput[i] =  0.0;
+            ptrInput[i] = 0.0;
+        }
+    }
+
+    Mem_Ref<CPU> get_input_memory()
+    {
+        return input;
+        
+    }
+
+    Mem_Ref<CPU> get_output_memory()
+    {
+        return output;
+        
+    }
+    void compute(Mem_Ref<CPU> input, Mem_Ref<CPU> output)
+    {
+        pybind11::buffer_info bufInput = input.request();
+        double *ptrInput = static_cast<double*>(bufInput.ptr);
+
+        pybind11::buffer_info bufOutput = output.request();
+        double *ptrOutput = static_cast<double*>(bufOutput.ptr);
+
+        for(int i = 0; i < size; i++)
+        {
+            ptrOutput[i] =  ptrInput[i];
+        }
+    }
+};
+
+template<>
+class Algo<CUDAGPU> {
+public:
+    Mem_Ref<CUDAGPU> input;
+    Mem_Ref<CUDAGPU> output;
+
+    int size;
+
+    void whoami()
+    {
+        std::cout << "I'm the GPU version\n";
+    }
+
+    void initialize_array(int size)
+    {
+        this->size = size;
+
+        pybind11::object input_allocate = cupy_allocate<double>({size});
```

Comment:
Actual, this should not working. I'm not sure, why it is working. You create a local variable, which holds the memory, create a reference off it, hold the reference and throw away the gpu memory after leaving the function. Actual the reference should become invalid, because it points to non existing memory. I think the python runtime is responsible for it, that the memory is not deleted.

Please store the `pybind11::object` instead the reference in the this class.

Reply:
so `Mem_Ref<CUDAGPU>` will be replaced by `py::object`? 

also do get_input() and get_ouput() need to be changed to `py::object` as the return type?

Reply:
The member variable of the `Algo` should be a `py::object` but the functions `get_input()` and `get_output()` should return an `Men_Ref`, because we don't want to give away the ownership of the memory.